### ISSUE TITLE
Writing NOTICE files for apps with the same name overwrites existing files

### ISSUE THREAD

**Issue by JanMattner** [github/licensed#371]
In large projects, one usually uses glob patterns to configure the source_paths.
Furthermore, one will have subprojects with the same name, such as "common" or "service".

When writing NOTICE files, licensed will just append the subproject name to the NOTICE file and overwrites any existing file. Thus, only the notices of the last subproject will remain.

Example:
Assume 3 subprojects "Service", the first having a dependency "A", the second "B" and the third "C".
When running `licensed notices`, I see this output:
```
Writing notices for MyTest-Service to C:/TEST/.licenses/NOTICE.MyTest-Service
Writing notices for MyTest-Service to C:/TEST/.licenses/NOTICE.MyTest-Service 
Writing notices for MyTest-Service to C:/TEST/.licenses/NOTICE.MyTest-Service
```
But in the file `NOTICE.MyTest-Service`, I only find the notice of Dependency "C" of the last Service subproject.

Better: instead of overwriting existing NOTICE files, just append the text.

Another idea: add the possibility to define a single notice file that is used for all apps. That makes sense, especially when using a shared cache anyway.

**Comment by jonabc** [comment]
There's a few things going on here, but generally I think without knowing more about your underlying use case I'm leaning towards not making any changes here.  I'd definitely welcome more context and information about your needs to better understand if there is really an issue here.

> Assume 3 subprojects "Service"

This straightaway seems like a problem.  Why are there three configured applications that all have the same name?  That seems like an anti-pattern in itself.

> Better: instead of overwriting existing NOTICE files, just append the text.

I don't think this is a better idea TBH, as it can lead to unintentionally mixing/matching/combining dependencies across configured applications. The notice file is not meant to display the dependencies for an entire repo, only for a single distributed application.  Licensed doesn't know the distribution model for the applications, services, libraries, or other distributed artifacts in a repository, and adding that sounds like it would add a level of undesired complexity.

> add the possibility to define a single notice file that is used for all apps. That makes sense, especially when using a shared cache anyway.

Same feedback here. Licensed doesn't know the distribution model for the artifacts created from a repo and makes notices for individual configured applications that can be distributed with that artifact.

Closing for now, but happy to keep discussing 👍

**Comment by JanMattner** [comment]
@jonabc I agree that apps / projects with the same name are an anti-pattern, but the problem here is, that for licensed, an "app" in a .NET project with NuGet dependency is just the top folder right below the "obj" folder (that is typically where the "*.csproj" file is located). But that top folder is not necessarily the name of the C# project resp. the output binary (usually a DLL or EXE).

Example:
In this repo: https://github.com/microsoft/VFSForGit/tree/master/GVFS
we have two projects ending in ".Windows":
GVFS.Installer.Windows and GVFS.Platform.Windows

They just followed an internal convention to put all project folders flatly next to each other. 
But in other projects, it is completely normal to group related projects in separate folders. But then the top folders of _two different projects_ within the _same app_ have the same name and licensed will just mix them up:
Alternative folder structure:
```
- GVFS
-- Installer
--- Windows
---- GVFS.Installer.Windows.csproj
-- Platform
--- Windows
---- GVFS.Platform.Windows.csproj
```
=> the namespace and DLLs of above projects are still different, but the top folders have the same names!

In such projects, licensed can only be used if you explicitly configure the "app" source_paths without globbing. But globbing is good for e.g. large scale repositories (100+ csproj "projects") or if you want to ensure that added projects cannot be forgotten in the license check.

Do you see any other feasible way to solve that issue with the current implementation?

**Comment by jonabc** [comment]
I hear you, but you could just as easily have grouped the above as

```
- GVFS
-- Windows
--- Installer
---- GVFS.Installer.Windows.csproj
--- Platform
---- GVFS.Platform.Windows.csproj
```

Either way, I'd expect that any one of those projects when built for distribution would only be reporting on the dependencies included in the build results.

Also, licensed does make this easier already if you specify a `name` property in conjunction with a glob pattern.  In the example added ☝️ , a config like the following would help
```yaml
apps:
  - source_path: Windows/*
     name: Windows # results in full names like Windows-Installer, Windows-Platform
  - source_path: Linux/*
     name: Linux # results in full names like Linux-Installer, Linux-Platform
```

In your original example, you could switch it around a bit:
```yaml
apps:
   - source_path: Installer/*
      name: Installer # results in full names like Installer-Windows
   - source_path: Platform/*
      name: Platform # results in full names like Platform-Windows
...
```

That's not as easy to use or generic as `**/*`, but now we're talking about how to make naming projects easier rather than combining notices from multiple projects together.

**Comment by JanMattner** [comment]
Yes, the problem is either managing the "app" names (currently: just the top folder) or the overwriting of NOTICE files. I just thought fixing the NOTICE files would be the easiest way.

Sure, your examples are already possible with the current means. That probably works in most cases for you, especially in smaller projects on GitHub. However, in the large scale enterprise world, things are different ;-)
e.g. project members won't be happy to restructure their 500+ projects solution (and yes, there is only one deliverable in the end, so it makes sense that this all together shares one NOTICE file) just for the sake of easier tooling.
The use case of "not forgetting added projects by using generic globbing" is also not covered currently (using generic globbing, the NOTICE files might be overwritten if the same top level folder name is used by accident; using your proposal with more specific naming scheme, new projects beyond those definitions won't be even found).

In the end, the core point is how easy should it be for a project to use licensed? If we reached here the end of flexibility, that's completely legit. If you do not see the need, why spending effort.
We'll still use licensed, it is a great tool, but we'll need to add some workarounds to overcome the remaining weaknesses of licensed.

So, too bad that this problem is not tackled, but we'll find a workaround.
From my side, the issue is also closed then and does not need to be reopened.

**Comment by jonabc** [comment]
@JanMattner please don't misunderstand, I'm not trying to come back with a message of "nothing will be changed to make your use case easier".  I misread the issue OP as asking specifically for a solution that alters the contents of the NOTICE files, which I don't believe is the best option.  If there are ways to better differentiate products by their names, great 👍 .

It gets increasingly tough to support every possible file structure with a generic "**/*" glob.  It sounds like you're using the full relative path from the root to a project folder as it's unique identifier.  Is that correct?  If so, that seems like a relatively easy configuration option to change auto-generated names from `project` to `path.from.root.to.project` or `path-from-root-to-project` or similar.  That might mean an overly long configured project name though.  WDYT?

> That probably works in most cases for you, especially in smaller projects on GitHub. However, in the large scale enterprise world, things are different ;-)
e.g. project members won't be happy to restructure their 500+ projects solution (and yes, there is only one deliverable in the end, so it makes sense that this all together shares one NOTICE file) just for the sake of easier tooling.
The use case of "not forgetting added projects by using generic globbing" is also not covered currently (using generic globbing, the NOTICE files might be overwritten if the same top level folder name is used by accident; using your proposal with more specific naming scheme, new projects beyond those definitions won't be even found).

I'd appreciate if you could avoid this tone of voice in the future.  We're all adults here.  I apologize if my tone came across insensitive on the previous message and I'll try to use a gentler tone in the future.  I'd appreciate you not making assumptions about the use cases, preferences or experiences that I or other users represent.

**Comment by JanMattner** [comment]
@jonabc I think there is a great misunderstanding here. I read again your and my posts and I do not see them in a harassing or provocative way. I also read again your code of conduct and I think we both stuck to it. Maybe as a non-native speaker it is more difficult for me to "strike the right chord" (not sure if that translation is correct), especially in written discussions only.
So I apologize for any inappropriate language, harassment, assumptions. So from my point of view you do not need to use a gentler tone, I am fine with it. I'll try to do it the same way.

I am happy to use that tool and your work / contributions, especially because I am not able to contribute except testing and giving feedback. That is exactly why I mean what I have written: if you do not see the need, I am completely fine with it. I have no "right" to insist on anything here.

About the assumption of use cases: apparantly, I am the first to see the need for this use case. So I assume, that no one else really tried to apply licensed to a similar large scale brown-field project (or they tried and didn't give feedback)
Of course this does not mean that you are other users only do "small projects on GitHub" and furthermore this does not rate in any way one type of project above another type.
I just try to give insights from another point of view, not sure if that is desired - again, if I am the only one, I am fine with not changing the tool just for me / my customers.


Back to the technical issue: Yes, in the end, I try to use a generic glob from the source root and just exclude unwanted paths/projects (e.g. test projects). I need to think about your proposal. I'll come back with feedback.

**Comment by JanMattner** [comment]
Sorry for the delay. 
I think in the end something like your proposal of a dynamic project name resolved from the root is the only feasible solution. Of course, names get long that way, but the user always has the option not to use generic globs and do detailed configuration instead. I think that's fine. 

An alternative would be e.g. creating GUIDs and a mapping file. But then you still have the problem of long names when you resolve the GUIDs for logging purpose - and if you do not resolve them, then the user has no idea which GUID is which project when checking the status logs... so better stick to speaking names. Easier to implement anyway. 

The folder separator in the project name cannot be a forbidden character in folder/ file names like the forward slash, because they are used e.g. as file name for the notice files. So it would be great if the separator is configurable, defaulting to a dash "-", that character is rarely used in a solution structure. 

What do you think?

**Comment by jonabc** [comment]
@JanMattner I took a look at code changes for relative path name generation and have everything working locally.  The PR will be a little bit longer to put up - I'm using the change as a forcing function to broadly refresh the configuration file documentation as well since that hasn't gotten enough attention recently.

> An alternative would be e.g. creating GUIDs and a mapping file. But then you still have the problem of long names when you resolve the GUIDs for logging purpose - and if you do not resolve them, then the user has no idea which GUID is which project when checking the status logs... so better stick to speaking names. Easier to implement anyway.

I hadn't considered this before but I agree with your conclusion - at some point you'll need to map everything back to names so the GUID conversion may not be worth it.

> The folder separator in the project name cannot be a forbidden character in folder/ file names like the forward slash, because they are used e.g. as file name for the notice files. So it would be great if the separator is configurable, defaulting to a dash "-", that character is rarely used in a solution structure.

👍 good idea, I made the separator configurable and defaulting to "-".  The only feedback here is that using the directory separator character in generated app names isn't an issue, you'll just end up recreating the relative directories to the source path inside the cache structure.  The app name is primarily used for organization and display and there isn't any need from the code perspective that the app name must be entirely represented in a single file/folder name.

> Of course, names get long that way, but the user always has the option not to use generic globs and do detailed configuration instead. I think that's fine.

This got me thinking about the scenario, and I think I might have something that could help.  I added a `depth` configuration property as well, with 0 being `all relative path parts between root and source path` and positive numeral values being the number of relative path parts to use in the app name starting from the `source_path` folder.

That's a mouthful, as an example a depth of 2 on a relative path like `lib/licensed/sources` would be `licensed/sources`.  That could help reduce app name length if some early path parts aren't significant.

I'll ping you on the PR when I open it in the next day or two to get your feedback 🙏
